### PR TITLE
Ensure http proxy authentication is supplied in httpc calls

### PR DESCRIPTION
**Pull request by lbarratt**
This PR ensures that calls to `:httpc.request` using the `:hex` profile supply `proxy_auth` as a `http_option()`

My use case is deploying elixir apps to an internal deployment of Cloud Foundry behind a firewall, limiting external connections to proxies. Hopefully others will find this useful!

### CODE REVIEW

**Comment by ericmj** [comment]
The tests seem to fail, I have fixed one failure on master but the other seem related to this PR. Can you rebase against master and see why the tests fail?

**Comment by lbarratt** [comment]
Thanks for looking at this @ericmj - I'll take a look

File: lib/hex/utils.ex [hunk 1]
```diff
@@ -216,4 +216,53 @@ defmodule Hex.Utils do
...
+    do: proxy_auth(http_proxy)
+  defp proxy_auth(%URI{scheme: "https"}, _http_proxy, https_proxy),
+    do: proxy_auth(https_proxy)
+
+  defp proxy_auth(nil),
+    do: []
+  defp proxy_auth(%URI{userinfo: nil}),
+    do: []
+  defp proxy_auth(%URI{userinfo: auth}) do
+    [user, pass] = case auth |> String.split(":") do
```
Comment (lexmag) [inline lib/hex/utils.ex]: Needless piping, It'll be more readable without it.

**Comment by lbarratt** [comment]
Okay, so the integration tests were failing because the proxy tests were setting `:http_proxy` in `Hex.State`. This meant that any subsequent http requests were failing as requests were attempting to connect via a non-existent proxy. 

I've been struggling to figure out the best way to reliably reset the state after each test so I'm not breaking things in other tests. I had success with running the proxy tests with `async: false` and resetting the proxy configuration in `Hex.State`. I also attempted to unset the proxy config for `:httpc` but wasn't able to figure it out (I'm not actually sure it is possible?). 

What I ended up with was passing a `NoProxy` option so that proxy requests to `localhost` were ignored. Is there a better way to go about this?

File: lib/hex/utils.ex (hunk 1 above)
Comment (lbarratt) [inline lib/hex/utils.ex]: Definitely, thanks. I've removed this :+1:

File: lib/hex/utils.ex [hunk 2]
```diff
@@ -220,4 +220,53 @@ defmodule Hex.Utils do
...
+  defp proxy_auth(%URI{userinfo: auth}) do
+    [user, pass] = case String.split(auth, ":") do
+      [user, pass] ->
+        [user, pass]
+      [user] ->
+        [user, ""]
+    end |> Enum.map(&String.to_char_list/1)
+
+    [proxy_auth: {user, pass}]
+  end
```
Comment (ericmj) [inline lib/hex/utils.ex]: Why the change here? Was there an issue with the `destructure` approach? We should definitely keep `parts: 2` in the `String.split` call.

**Comment by ericmj** [comment]
Left one comment, rest looks great. Please squash into a single commit when you are done.

File: lib/hex/utils.ex (hunk 2 above)
Comment (lbarratt) [inline lib/hex/utils.ex]: I'm not sure exactly, I think I had been struggling with handling the possibility of `pass` being `nil` and then converting the values to char lists. Somehow I ended up with `case`.

Having revisited this I figured out a potentially better way using `destructure`:

```elixir
destructure [user, pass], String.split(auth, ":", parts: 2)
[user, pass] = Enum.map([user, pass || ""], &(String.to_char_list(&1)))

[proxy_auth: {user, pass}]
```

What do you think? if this is okay I'll commit this and squash them.

Comment (ericmj) [inline lib/hex/utils.ex]: Yes, although I would split the `Enum.map` into two separate explicit expressions.

Comment (lbarratt) [inline lib/hex/utils.ex]: Thanks again for taking a look! Apologies, I'm not sure what you mean by separating them explicitly?

I have pushed the suggested change in the meantime

File: lib/hex/utils.ex [hunk 3]
```diff
@@ -221,4 +221,49 @@ defmodule Hex.Utils do
...
+  defp proxy_auth(%URI{scheme: "https"}, _http_proxy, https_proxy),
+    do: proxy_auth(https_proxy)
+
+  defp proxy_auth(nil),
+    do: []
+  defp proxy_auth(%URI{userinfo: nil}),
+    do: []
+  defp proxy_auth(%URI{userinfo: auth}) do
+    destructure [user, pass], String.split(auth, ":", parts: 2)
+    [user, pass] = Enum.map([user, pass || ""], &String.to_char_list/1)
```
Comment (lexmag) [inline lib/hex/utils.ex]: I believe @ericmj has meant:
```elixir
user = String.to_char_list(user)
pass = String.to_char_list(pass || "")
```
I'd say it looks better than having `Enum.map` call with `||` inside first argument.

Comment (lbarratt) [inline lib/hex/utils.ex]: Ah, this makes sense. Thank you @lexmag

I've made the change and rebased on master. If this looks good i'll squash the commits

**Comment by ericmj** [comment]
@lbarratt Looks great, please squash.

**Comment by lbarratt** [comment]
@ericmj Squashed. Cheers!
